### PR TITLE
Multi tests

### PR DESCRIPTION
**Pull request by peaBerberian**
This is a collection of commits which:
  - allows to run multiple tests, potentially each with its own chart
  - automatically exports tests when a test is finished
  - sanitize inputs given to the back-end
  - separate the chart-displaying logic from the data-registering logic.
    This is both to allow registering data not linked to the chart (ex: amount of rebuffering) and to facilitate the data import/export logic
  - updating the three scripts `setup.sh`, `start_server.sh` and `start_client.sh` so those are the only script needed to use the tool
  - fix linting errors (according to its eslint configuration) that were in the back-end
  - update the toxics as a whole instead of one by one (for the moment, seems more adapted to what we want to do. To see if we want a better granularity in the future).

### CODE REVIEW

File: cdn-video-server/src/toxiproxy/utils.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
+import cp from "child_process";
+import path from "path";
+import {
+  IApiToxic,
+  INormalizedToxic,
+  IProxy,
+  ToxicType,
+} from "./toxiproxyTypes";
+
+// TODO Might be better written somewhere else
```
Comment (alaamoum) [inline cdn-video-server/src/toxiproxy/utils.ts:10]: I agree, there should be a const/utils file for all the const

File: cdn-video-server/src/toxiproxy/utils.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      const toxic = toxics[i];
+      const normalizedToxic = normalizeToxic(toxic);
+      const attributeList = dictToString(normalizedToxic.attributes);
+      const streamFlag = normalizedToxic.stream === "downstream" ? "-d" : "-u";
+      proms.push(execProm(
+        `${TOXIPROXY_CLI_EXEC} toxic add -t ${normalizedToxic.type} -a ${attributeList} --tox ${normalizedToxic.toxicity} ${streamFlag} -n ${normalizedToxic.name} ${this.name}`,
+      ));
+      this.toxics.push(normalizedToxic);
+    }
+    return Promise.all(proms).then(() => {});
```
Comment (alaamoum) [inline cdn-video-server/src/toxiproxy/utils.ts:60]: We could simplify this by doing : 
```typescript
toxics.map(normalizeToxic).forEach(toxic => ...) 
```

File: cdn-video-server/src/toxiproxy/utils.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    });
+  });
+}
+
+let id = 0;
+function generateToxicName() : string {
+  id += 1;
+  return String(id);
+}
+
```
Comment (alaamoum) [inline cdn-video-server/src/toxiproxy/utils.ts:90]: There is a default name for the toxics `{proxy}-{type_of_toxic}-{downstream/upstream}` if we don't specify a name, and I think they are more clear than just a integer id

File: cdn-video-server/src/toxiproxy/utils.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+function normalizeToxic(toxic: IApiToxic): INormalizedToxic {
+  checkNumberProperty(toxic.toxicity, "toxicity");
+  const toxicity = toxic.toxicity ?? 1;
+  if (
+    toxic.stream !== undefined
+    && toxic.stream !== "upstream"
+    && toxic.stream !== "downstream"
+  ) {
+    throw new Error("Invalid stream property given.");
```
Comment (alaamoum) [inline cdn-video-server/src/toxiproxy/utils.ts:99]: If we don't have a stram, it should default to `downstream` as per toxiproxy's behavior

File: cdn-video-server/src/toxiproxy/utils.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  }
+  const stream = toxic.stream ?? "downstream";
+
+  if (typeof toxic.type !== "string") {
+    throw new Error("Invalid type property given: not a string.");
+  }
```
Comment (alaamoum) [inline cdn-video-server/src/toxiproxy/utils.ts:105]: Maybe we should use an Enum for the toxic types and do `Object.values(EnumToxicType).includes(toxic.type)` to be more thorough

File: cdn-video-server/src/toxiproxy/utils.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  if (typeof toxic.attributes !== "object" || toxic.attributes === null) {
+    throw new Error("Invalid attributes property given");
+  }
```
Comment (alaamoum) [inline cdn-video-server/src/toxiproxy/utils.ts:108]: Maybe we should use some pattern matching to verify the attributes ( cf https://github.com/gvergnaud/ts-pattern )

File: player-tester/src/comparator-handler.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+document.getElementById("import").onclick = function() {
+    const files = document.getElementById("file").files;
+    console.log(files)
+    if (files.length <= 0) {
+        window.alert("No file selected")
+        return false;
+    }
+    const fileReader = new FileReader();
+    fileReader.onload = (event) => {
+        const data = JSON.parse(event.target.result)
```
Comment (alaamoum) [inline player-tester/src/comparator-handler.js:14]: Just a question, I didn't test this , but if we import two files, our chart will always refer to the first file ( as we don't remove the files from the document files, i.e `document.getElementById("file").files` will always have 2 files, and `event.target.result` will always give the same file descriptor no ?

**Review by achrafl0 — APPROVED** [review]

File: cdn-video-server/src/toxiproxy/utils.ts (hunk 3 above)
Comment (peaBerberian) [reply cdn-video-server/src/toxiproxy/utils.ts:90]: Yes but we need to know the name to be able to delete them. We could perhaps add a more type-related name but that would be more complicated for few gain.

The README.md of toxiproxy indicates how they format the name by default but I don't know to which extents it's actually part of their API, so I preferred not to rely on that and preferred defining it myself.

File: cdn-video-server/src/toxiproxy/utils.ts (hunk 4 above)
Comment (peaBerberian) [reply cdn-video-server/src/toxiproxy/utils.ts:99]: Yes, that's what I implemented here.

File: cdn-video-server/src/toxiproxy/utils.ts (hunk 6 above)
Comment (peaBerberian) [reply cdn-video-server/src/toxiproxy/utils.ts:108]: I don't know about that.
This check is straightforward and idiomatic. And it also gives an explicit error of the problem back to the client.

I like pattern matching when included in a language, but I'm under the impression that in this precize case we would be adding a library with harder-to-understand syntax for no real advantage.

Comment (alaamoum) [reply cdn-video-server/src/toxiproxy/utils.ts:108]: The pattern matching was to check if we have the correct attributes for the toxic type

File: cdn-video-server/src/toxiproxy/utils.ts (hunk 2 above)
Comment (alaamoum) [reply cdn-video-server/src/toxiproxy/utils.ts:60]: It's in no way objective, I just like the map syntax better than the loop syntax ^^

**Pull request merged**